programs/survival/waystones.sc

```diff
@@ -0,0 +1,310 @@
+                'name'->_get_name_from_nbt(item:2) || floor+' Waypoint' 
+            };
+            _app_message(p, str('Activated %s Waystone', global_waystones:pos:'name'));
+            _write_waystones();
+            sound( 'block.enchantment_table.use', pos );
+            sound( 'minecraft:block.amethyst_block.hit', pos );
+        ,
+            _app_message(p, str('Waystone structure requires %d %s to activate',
+                global_settings:'structure_size',global_settings:'structure_material'));
+            sound( 'block.copper.step', pos );
+        );
+    );
+);
+
+__on_player_breaks_block(p, block)->(
+    if(block~'lodestone',
+        pos = pos(block);
+        if(global_waystones:pos,
+            sound( 'minecraft:item.trident.thunder', pos );
+            _app_message(p, str('Removed %s Waystone', global_waystones:pos:'name'));
+            delete(global_waystones:pos); 
+        );
+    );
+);
+
+__on_player_right_clicks_block(p, item, hand, block, face, hitvec) -> (
+    if(block~'lodestone' && !p~'sneaking',
+        dimension = current_dimension();
+        if(global_settings:'enabled_dimensions'~dimension,
+            pos = pos(block);
+            
+            if(global_waystones:pos,
+                if( !_handle_name_tag(item, pos, p, hand) && hand=='mainhand',
+                    uuid = p~'uuid';
+                    _read_player_waypoints(uuid);
+                    _mark_player_waypoint(pos, p~'pos', uuid); 
+                    sound( 'minecraft:block.amethyst_block.hit', block );
+                    _open_waypoint_screen(p, pos, uuid);
+                    _write_player_waypoints(uuid);
+                );
+            ,
+                _app_message(p, str('Place the lodestone near a %d block %s structure to activate',
+                    global_settings:'structure_size',global_settings:'structure_material'));
+            );
+        ,
+            _app_message(p, 'Waystones dont\'t work in this dimension');
+            sound( 'item.trident.return', block );
+        );
+    );
+);
+
+_has_required_structure(pos) -> (
+    if(global_settings:'structure_material' && global_settings:'structure_size', 
+        count = 0;
+        scan(pos, 2,2,2, if( _~(global_settings:'structure_material'), count+=1) );
+        count >= global_settings:'structure_size';
+    ,true);
+);
+
+_has_required_item(p) -> (
+    if(!global_settings:'offering', return(true));
+    slot = inventory_find(p, global_settings:'offering');
+    if(slot,
+        item = inventory_get(p, slot);
+        inventory_set(p, slot, item:1 - 1);
+        true;
+    ,
+        _app_message(p, str('The Waystone requires an %s offering', global_settings:'offering'));
+        sound( 'item.trident.return', p~'pos' );
+        false;
+    );
+);
+
+_warp_player(p, pos, dimension, point)->(
+    if(_has_required_item(p),
+        in_dimension(dimension, 
+            run(str('tp %s %d %d %d', p~'command_name', ...(pos + [0.5, 0, 0.5]) ));
```

Comment:
Small thing: The player will suffocate if the waystone spawn has been covered! Is it worth it to have logic to find a safe place?

Reply:
Probably not in my opinion. My current solution saves the last location they opened the way stone from. That should be safe, aside from another player filling in the area with blocks. If this was used on a public server with poorly behaved players that could be an issue. It could also be a amusing prank, similar to filling in the area a player logged out from, or boxing in a afk player.

Finding a safe location would probably involve either randomly picking block around the stone or looping over the area around the stone. I think that's a lot of code for something that's not going to happen very often.